File: sources/Android/android-communications/library/src/main/java/com/polar/androidcommunications/api/ble/model/gatt/client/pmd/model/PpgData.kt
package com.polar.androidcommunications.api.ble.model.gatt.client.pmd.model

import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.BlePMDClient
import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.PmdDataFrame
import com.polar.androidcommunications.api.ble.model.gatt.client.pmd.PmdDataFrameUtils
import com.polar.androidcommunications.common.ble.TypeUtils
import kotlin.experimental.and

/**
 * Sealed class to represent Ppg data sample
 */
internal sealed class PpgDataSample

internal class PpgData {
    // PPG Data Sample 0
    data class PpgDataFrameType0 internal constructor(
        val timeStamp: ULong,
        val ppgDataSamples: List<Int>,
        val ambientSample: Int
    ) : PpgDataSample()

    // PPG Data Sample 3
    data class PpgDataFrameType8 internal constructor(
        val timeStamp: ULong,
        val ppgDataSamples: List<Int>,
        val status: UInt
    ) : PpgDataSample()

    // PPG Data frame type 4
    data class PpgDataFrameType4 internal constructor(
        val timeStamp: ULong,
        val numIntTs: List<UInt>,
        val channel1GainTs: List<UInt>,
        val channel2GainTs: List<UInt>
    ) : PpgDataSample()

    // PPG Data frame type 5
    data class PpgDataFrameType5 internal constructor(
        val timeStamp: ULong,
        val operationMode: UInt
    ) : PpgDataSample()

    // PPG Data Sample 2
    data class PpgDataFrameType7 internal constructor(
        val timeStamp: ULong,
        val ppgDataSamples: List<Int>,
        val status: UInt
    ) : PpgDataSample()

    data class PpgDataFrameType9 internal constructor(
        val timeStamp: ULong,
        val numIntTs: List<UInt>,
        val channel1GainTs: List<UInt>,
        val channel2GainTs: List<UInt>
    ) : PpgDataSample()

    data class PpgDataFrameType10 internal constructor(
        val timeStamp: ULong,
        val greenSamples: List<Int>,
        val redSamples: List<Int>,
        val irSamples: List<Int>,
        val status: Int
    ) : PpgDataSample()

    // PPG Data Sport Id
    data class PpgDataSampleSportId internal constructor(
        val timeStamp: ULong,
        val sportId: ULong
    ) : PpgDataSample()

    val ppgSamples: MutableList<PpgDataSample> = ArrayList()

    companion object {
        private const val TAG = "PpgData"

        private const val TYPE_0_SAMPLE_SIZE_IN_BYTES = 3
        private const val TYPE_0_SAMPLE_SIZE_IN_BITS = TYPE_0_SAMPLE_SIZE_IN_BYTES * 8
        private const val TYPE_0_CHANNELS_IN_SAMPLE = 4

        private const val TYPE_4_NUM_INTS_SIZE = 12
        private const val TYPE_4_CHANNEL_0_AND_1_SIZE = 24
        private const val TYPE_4_SAMPLE_SIZE_IN_BYTES =
            TYPE_4_NUM_INTS_SIZE + TYPE_4_CHANNEL_0_AND_1_SIZE

        private const val TYPE_5_SAMPLE_SIZE_IN_BYTES = 4

        private const val TYPE_6_SAMPLE_SIZE_IN_BYTES = 8

        private const val TYPE_7_SAMPLE_SIZE_IN_BYTES = 3
        private const val TYPE_7_SAMPLE_SIZE_IN_BITS = TYPE_7_SAMPLE_SIZE_IN_BYTES * 8
        private const val TYPE_7_CHANNELS_IN_SAMPLE = 17

        private const val TYPE_8_SAMPLE_SIZE_IN_BYTES = 3
        private const val TYPE_8_SAMPLE_SIZE_IN_BITS = TYPE_8_SAMPLE_SIZE_IN_BYTES * 8
        private const val TYPE_8_CHANNELS_IN_SAMPLE = 25

        private const val TYPE_9_NUM_INTS_SIZE = 12
        private const val TYPE_9_CHANNEL_0_AND_1_SIZE = 24
        private const val TYPE_9_SAMPLE_SIZE_IN_BYTES =
            TYPE_9_NUM_INTS_SIZE + TYPE_9_CHANNEL_0_AND_1_SIZE

        private const val TYPE_10_SAMPLE_SIZE_IN_BYTES = 3
        private const val TYPE_10_SAMPLE_SIZE_IN_BITS = TYPE_10_SAMPLE_SIZE_IN_BYTES * 8
        private const val TYPE_10_STATUS_SIZE = 20
        private const val TYPE_10_CHANNELS_IN_SAMPLE = 21

        fun parseDataFromDataFrame(frame: PmdDataFrame): PpgData {
            return if (frame.isCompressedFrame) {
                when (frame.frameType) {
                    PmdDataFrame.PmdDataFrameType.TYPE_0 -> dataFromCompressedType0(frame)
                    PmdDataFrame.PmdDataFrameType.TYPE_7 -> dataFromCompressedType7(frame)
                    PmdDataFrame.PmdDataFrameType.TYPE_8 -> dataFromCompressedType8(frame)
                    PmdDataFrame.PmdDataFrameType.TYPE_10 -> dataFromCompressedType10(frame)
                    else -> throw java.lang.Exception("Compressed FrameType: ${frame.frameType} is not supported by PPG data parser")
                }
            } else {
                when (frame.frameType) {
                    PmdDataFrame.PmdDataFrameType.TYPE_0 -> dataFromRawType0(frame)
                    PmdDataFrame.PmdDataFrameType.TYPE_4 -> dataFromRawType4(frame)
                    PmdDataFrame.PmdDataFrameType.TYPE_5 -> dataFromRawType5(frame)
                    PmdDataFrame.PmdDataFrameType.TYPE_6 -> dataFromRawType6(frame)
                    PmdDataFrame.PmdDataFrameType.TYPE_9 -> dataFromRawType9(frame)
                    else -> throw java.lang.Exception("Raw FrameType: ${frame.frameType} is not supported by PPG data parser")
                }
            }
        }

        private fun dataFromRawType0(frame: PmdDataFrame): PpgData {
            val ppgData = PpgData()
            val step = TYPE_0_SAMPLE_SIZE_IN_BYTES

            val samplesSize = frame.dataContent.size / (step * TYPE_0_CHANNELS_IN_SAMPLE)
            val timeStamps = PmdTimeStampUtils.getTimeStamps(
                previousFrameTimeStamp = frame.previousTimeStamp,
                frameTimeStamp = frame.timeStamp,
                samplesSize = samplesSize,
                frame.sampleRate
            )
            var timeStampIndex = 0

            var i = 0
            while (i < frame.dataContent.size) {
                val samples: MutableList<Int> = mutableListOf()
                for (ch in 0 until TYPE_0_CHANNELS_IN_SAMPLE) {
                    samples.add(TypeUtils.convertArrayToSignedInt(frame.dataContent, i, step))
                    i += step
                }
                ppgData.ppgSamples.add(
                    PpgDataFrameType0(
                        timeStamp = timeStamps[timeStampIndex],
                        samples.subList(0, 3),
                        samples[3]
                    )
                )
                timeStampIndex++
            }
            return ppgData
        }

        private fun dataFromRawType4(frame: PmdDataFrame): PpgData {
            val ppgData = PpgData()

            val samplesSize = frame.dataContent.size / TYPE_4_SAMPLE_SIZE_IN_BYTES
            val timeStamps = PmdTimeStampUtils.getTimeStamps(
                previousFrameTimeStamp = frame.previousTimeStamp,
                frameTimeStamp = frame.timeStamp,
                samplesSize = samplesSize,
                frame.sampleRate
            )
            var timeStampIndex = 0

            var offset = 0
            while (offset < frame.dataContent.size) {
                val numIntTs =
                    frame.dataContent.sliceArray(offset until (offset + TYPE_4_NUM_INTS_SIZE))
                        .map { it.toUByte().toUInt() }
                offset += TYPE_4_NUM_INTS_SIZE
                val channel1GainTs =
                    frame.dataContent.sliceArray(offset until (offset + TYPE_4_CHANNEL_0_AND_1_SIZE))
                        .toList()
                        .mapIndexedNotNull { idx, v -> if (idx % 2 == 0) (v and 0x07).toUInt() else null }
                val channel2GainTs =
                    frame.dataContent.sliceArray(offset until (offset + TYPE_4_CHANNEL_0_AND_1_SIZE))
                        .toList()
                        .mapIndexedNotNull { idx, v -> if (idx % 2 == 1) (v and 0x07).toUInt() else null }
                offset += TYPE_4_CHANNEL_0_AND_1_SIZE
                ppgData.ppgSamples.add(
                    PpgDataFrameType4(
                        timeStamp = timeStamps[timeStampIndex],
                        numIntTs = numIntTs,
                        channel1GainTs = channel1GainTs,
                        channel2GainTs = channel2GainTs
                    )
                )
                timeStampIndex++
            }
            return ppgData
        }

        private fun dataFromRawType5(frame: PmdDataFrame): PpgData {
            val ppgData = PpgData()
            var offset = 0

            val samplesSize = frame.dataContent.size / TYPE_5_SAMPLE_SIZE_IN_BYTES
            val timeStamps = PmdTimeStampUtils.getTimeStamps(
                previousFrameTimeStamp = frame.previousTimeStamp,
                frameTimeStamp = frame.timeStamp,
                samplesSize = samplesSize,
                frame.sampleRate
            )
            var timeStampIndex = 0

            while (offset < frame.dataContent.size) {
                val operationMode = PmdDataFrameUtils.parseFrameDataField(
                    frame.dataContent.sliceArray(offset until (offset + TYPE_5_SAMPLE_SIZE_IN_BYTES)),
                    BlePMDClient.PmdDataFieldEncoding.UNSIGNED_INT
                ) as UInt
                offset += TYPE_5_SAMPLE_SIZE_IN_BYTES
                ppgData.ppgSamples.add(
                    PpgDataFrameType5(
                        timeStamp = timeStamps[timeStampIndex],
                        operationMode = operationMode
                    )
                )
                timeStampIndex++
            }
            return ppgData
        }

        private fun dataFromRawType6(frame: PmdDataFrame): PpgData {
            val ppgData = PpgData()
            val sportId = PmdDataFrameUtils.parseFrameDataField(
                frame.dataContent.sliceArray(0 until TYPE_6_SAMPLE_SIZE_IN_BYTES),
                BlePMDClient.PmdDataFieldEncoding.UNSIGNED_LONG
            ) as ULong

            val samplesSize = frame.dataContent.size / TYPE_6_SAMPLE_SIZE_IN_BYTES
            val timeStamps = PmdTimeStampUtils.getTimeStamps(
                previousFrameTimeStamp = frame.previousTimeStamp,
                frameTimeStamp = frame.timeStamp,
                samplesSize = samplesSize,
                frame.sampleRate
            )

            ppgData.ppgSamples.add(
                PpgDataSampleSportId(
                    timeStamp = timeStamps.first(),
                    sportId = sportId
                )
            )
            return ppgData
        }

        private fun dataFromRawType9(frame: PmdDataFrame): PpgData {
            val ppgData = PpgData()

            val samplesSize = frame.dataContent.size / TYPE_9_SAMPLE_SIZE_IN_BYTES
            val timeStamps = PmdTimeStampUtils.getTimeStamps(
                previousFrameTimeStamp = frame.previousTimeStamp,
                frameTimeStamp = frame.timeStamp,
                samplesSize = samplesSize,
                frame.sampleRate
            )
            var timeStampIndex = 0

            var offset = 0
            while (offset < frame.dataContent.size) {
                val numIntTs =
                    frame.dataContent.sliceArray(offset until (offset + TYPE_9_NUM_INTS_SIZE))
                        .map { it.toUByte().toUInt() }
                offset += TYPE_9_NUM_INTS_SIZE
                val channel1GainTs =
                    frame.dataContent.sliceArray(offset until (offset + TYPE_9_CHANNEL_0_AND_1_SIZE))
                        .toList()
                        .mapIndexedNotNull { idx, v -> if (idx % 2 == 0) (v and 0x07).toUInt() else null }
                val channel2GainTs =
                    frame.dataContent.sliceArray(offset until (offset + TYPE_9_CHANNEL_0_AND_1_SIZE))
                        .toList()
                        .mapIndexedNotNull { idx, v -> if (idx % 2 == 1) (v and 0x07).toUInt() else null }
                offset += TYPE_9_CHANNEL_0_AND_1_SIZE
                ppgData.ppgSamples.add(
                    PpgDataFrameType9(
                        timeStamp = timeStamps[timeStampIndex],
                        numIntTs = numIntTs,
                        channel1GainTs = channel1GainTs,
                        channel2GainTs = channel2GainTs
                    )
                )
                timeStampIndex++
            }
            return ppgData
        }

        private fun dataFromRawType10(frame: PmdDataFrame): PpgData {
            val ppgData = PpgData()

            val samplesSize = frame.dataContent.size / TYPE_10_SAMPLE_SIZE_IN_BYTES
            val timeStamps = PmdTimeStampUtils.getTimeStamps(
                previousFrameTimeStamp = frame.previousTimeStamp,
                frameTimeStamp = frame.timeStamp,
                samplesSize = samplesSize,
                frame.sampleRate
            )
            var timeStampIndex = 0

            var offset = 0
            while (offset < frame.dataContent.size) {
                val grn =
                    frame.dataContent.sliceArray(offset until (offset + TYPE_10_SAMPLE_SIZE_IN_BYTES))
                        .map { it.toUByte().toInt() }
                offset += TYPE_10_SAMPLE_SIZE_IN_BYTES
                val red =
                    frame.dataContent.sliceArray(offset until (offset + TYPE_10_SAMPLE_SIZE_IN_BYTES))
                        .map { it.toUByte().toInt() }
                offset += TYPE_10_SAMPLE_SIZE_IN_BYTES
                val ir =
                    frame.dataContent.sliceArray(offset until (offset + TYPE_10_SAMPLE_SIZE_IN_BYTES))
                        .map { it.toUByte().toInt() }
                offset += TYPE_10_SAMPLE_SIZE_IN_BYTES
                val status =
                    frame.dataContent.sliceArray(offset until (offset + TYPE_10_STATUS_SIZE))
                        .map { it.toUByte().toInt() }.last()
                offset += TYPE_10_STATUS_SIZE

                ppgData.ppgSamples.add(
                    PpgDataFrameType10(
                        timeStamp = timeStamps[timeStampIndex],
                        redSamples = red,
                        greenSamples = grn,
                        irSamples = ir,
                        status = status
                    )
                )
                timeStampIndex++
            }
            return ppgData
        }

        private fun dataFromCompressedType0(frame: PmdDataFrame): PpgData {
            val ppgData = PpgData()
            val samples = BlePMDClient.parseDeltaFramesAll(
                frame.dataContent,
                TYPE_0_CHANNELS_IN_SAMPLE,
                TYPE_0_SAMPLE_SIZE_IN_BITS,
                BlePMDClient.PmdDataFieldEncoding.SIGNED_INT
            )

            val timeStamps = PmdTimeStampUtils.getTimeStamps(
                previousFrameTimeStamp = frame.previousTimeStamp,
                frameTimeStamp = frame.timeStamp,
                samplesSize = samples.size,
                frame.sampleRate
            )

            for ((index, sample) in samples.withIndex()) {
                val ppg0 = (sample[0].toFloat() * frame.factor).toInt()
                val ppg1 = (sample[1].toFloat() * frame.factor).toInt()
                val ppg2 = (sample[2].toFloat() * frame.factor).toInt()
                val ambient = (sample[3].toFloat() * frame.factor).toInt()
                ppgData.ppgSamples.add(
                    PpgDataFrameType0(
                        timeStamp = timeStamps[index],
                        ppgDataSamples = listOf(ppg0, ppg1, ppg2),
                        ambient
                    )
                )
            }
            return ppgData
        }

        private fun dataFromCompressedType7(frame: PmdDataFrame): PpgData {
            val ppgData = PpgData()
            val samples = BlePMDClient.parseDeltaFramesAll(
                frame.dataContent,
                TYPE_7_CHANNELS_IN_SAMPLE,
                TYPE_7_SAMPLE_SIZE_IN_BITS,
                BlePMDClient.PmdDataFieldEncoding.SIGNED_INT
            )

            val timeStamps = PmdTimeStampUtils.getTimeStamps(
                previousFrameTimeStamp = frame.previousTimeStamp,
                frameTimeStamp = frame.timeStamp,
                samplesSize = samples.size,
                frame.sampleRate
            )

            for ((index, sample) in samples.withIndex()) {
                val channels = sample.subList(0, 16).map {
                    if (frame.factor != 1.0f) (it.toFloat() * frame.factor).toInt() else it
                }
                val status = (sample[16] and 0xFFFFFF).toUInt()
                ppgData.ppgSamples.add(
                    PpgDataFrameType7(
                        timeStamp = timeStamps[index],
                        ppgDataSamples = channels,
                        status
                    )
                )
            }
            return ppgData
        }

        private fun dataFromCompressedType8(frame: PmdDataFrame): PpgData {
            val ppgData = PpgData()
            val samples = BlePMDClient.parseDeltaFramesAll(
                frame.dataContent,
                TYPE_8_CHANNELS_IN_SAMPLE,
                TYPE_8_SAMPLE_SIZE_IN_BITS,
                BlePMDClient.PmdDataFieldEncoding.SIGNED_INT
            )

            val timeStamps = PmdTimeStampUtils.getTimeStamps(
                previousFrameTimeStamp = frame.previousTimeStamp,
                frameTimeStamp = frame.timeStamp,
                samplesSize = samples.size,
                frame.sampleRate
            )

            for ((index, sample) in samples.withIndex()) {
                val channels = sample.subList(0, 24).map {
                    if (frame.factor != 1.0f) (it.toFloat() * frame.factor).toInt() else it
                }
                val status = (sample[24] and 0xFFFFFF).toUInt()
                ppgData.ppgSamples.add(
                    PpgDataFrameType8(
                        timeStamp = timeStamps[index],
                        ppgDataSamples = channels,
                        status
                    )
                )
            }
            return ppgData
        }

        private fun dataFromCompressedType10(frame: PmdDataFrame): PpgData {
            val ppgData = PpgData()
            val samples = BlePMDClient.parseDeltaFramesAll(
                frame.dataContent,
                TYPE_10_CHANNELS_IN_SAMPLE,
                TYPE_10_SAMPLE_SIZE_IN_BITS,
                BlePMDClient.PmdDataFieldEncoding.SIGNED_INT
            )

            val timeStamps = PmdTimeStampUtils.getTimeStamps(
                previousFrameTimeStamp = frame.previousTimeStamp,
                frameTimeStamp = frame.timeStamp,
                samplesSize = samples.size,
                frame.sampleRate
            )

            for ((index, sample) in samples.withIndex()) {
                val greenSamples = sample.subList(0, 8).map {
                    if (frame.factor != 1.0f) (it.toFloat() * frame.factor).toInt() else it
                }

                val redSamples = sample.subList(8, 14).map {
                    if (frame.factor != 1.0f) (it.toFloat() * frame.factor).toInt() else it
                }

                val irSamples = sample.subList(14, 20).map {
                    if (frame.factor != 1.0f) (it.toFloat() * frame.factor).toInt() else it
                }

                val status = sample.last()

                ppgData.ppgSamples.add(
                    PpgDataFrameType10(
                        timeStamp = timeStamps[index],
                        greenSamples = greenSamples,
                        redSamples = redSamples,
                        irSamples = irSamples,
                        status
                    )
                )
            }
            return ppgData
        }
    }
}

File: sources/iOS/ios-communications/Sources/iOSCommunications/ble/api/model/gatt/client/pmd/model/PpgData.swift
//  Copyright © 2022 Polar. All rights reserved.

import Foundation

public class PpgData {
    let timeStamp: UInt64
    
    struct PpgSample {
        let timeStamp: UInt64?
        let ppgDataSamples: [Int32]
        let ambientSample: Int32!
        let status: Int32!
        let frameType: PmdDataFrameType
    }
    
    var samples: [PpgSample]
    
    
    init(timeStamp: UInt64 = 0, samples: [PpgSample] = []) {
        self.timeStamp = timeStamp
        self.samples = samples
    }
    
    private static let TYPE_0_SAMPLE_SIZE_IN_BYTES: UInt8 = 3
    private static let TYPE_0_SAMPLE_SIZE_IN_BITS: UInt8 = TYPE_0_SAMPLE_SIZE_IN_BYTES * 8
    private static let TYPE_0_CHANNELS_IN_SAMPLE: UInt8 = 4
    private static let TYPE_6_SAMPLE_SIZE_IN_BYTES: Int = 8
    private static let TYPE_7_SAMPLE_SIZE_IN_BYTES: UInt8 = 3
    private static let TYPE_7_SAMPLE_SIZE_IN_BITS: UInt8 = TYPE_7_SAMPLE_SIZE_IN_BYTES * 8
    private static let TYPE_7_CHANNELS_IN_SAMPLE: UInt8 = 17
    private static let TYPE_9_NUM_INTS_SIZE = 12
    private static let TYPE_9_CHANNEL_0_AND_1_SIZE = 24
    private static let TYPE_9_SAMPLE_SIZE_IN_BYTES =
    TYPE_9_NUM_INTS_SIZE + TYPE_9_CHANNEL_0_AND_1_SIZE
    private static let TYPE_10_SAMPLE_SIZE_IN_BYTES: UInt8 = 3
    private static let TYPE_10_SAMPLE_SIZE_IN_BITS: UInt8 = TYPE_10_SAMPLE_SIZE_IN_BYTES * 8
    private static let TYPE_10_STATUS_SIZE: UInt8 = 20
    private static let TYPE_10_CHANNELS_IN_SAMPLE: UInt8 = 21
    
    static func parseDataFromDataFrame(frame: PmdDataFrame) throws -> PpgData {
        if (frame.isCompressedFrame) {
            switch (frame.frameType) {
            case PmdDataFrameType.type_0: return try dataFromCompressedType0(frame: frame)
            case PmdDataFrameType.type_7: return try dataFromCompressedType7(frame: frame)
            case PmdDataFrameType.type_10: return try dataFromCompressedType10(frame: frame)
            default: throw BleGattException.gattDataError(description: "Compressed FrameType: \(frame.frameType) is not supported by PPG data parser")
            }
        } else {
            switch (frame.frameType) {
            case PmdDataFrameType.type_0: return try dataFromRawType0(frame: frame)
            case PmdDataFrameType.type_6: return try dataFromRawType6(frame: frame)
            case PmdDataFrameType.type_9: return try dataFromRawType9(frame: frame)
            default: throw BleGattException.gattDataError(description: "Raw FrameType: \(frame.frameType) is not supported by PPG data parser")
            }
        }
    }
    
    private static func dataFromRawType0(frame: PmdDataFrame) throws -> PpgData {
        var offset = 0
        let step = TYPE_0_SAMPLE_SIZE_IN_BYTES
        let samplesSize = Int(Double(frame.dataContent.count) / Double(step * TYPE_0_CHANNELS_IN_SAMPLE))
        
        let timeStamps = try PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp: frame.previousTimeStamp, frameTimeStamp: frame.timeStamp, samplesSize: UInt(samplesSize), sampleRate: frame.sampleRate)

        var timeStampIndex = 0
        var ppgSamples = [PpgSample]()
        while offset < frame.dataContent.count {
            let ppg0 = TypeUtils.convertArrayToSignedInt(frame.dataContent, offset: offset, size: Int(step))
            offset += Int(step)
            let ppg1 = TypeUtils.convertArrayToSignedInt(frame.dataContent, offset: offset, size: Int(step))
            offset += Int(step)
            let ppg2 = TypeUtils.convertArrayToSignedInt(frame.dataContent, offset: offset, size: Int(step))
            offset += Int(step)
            let ambient = TypeUtils.convertArrayToSignedInt(frame.dataContent, offset: offset, size: Int(step))
            offset += Int(step)
            
            ppgSamples.append( PpgSample( timeStamp: timeStamps[timeStampIndex], ppgDataSamples: [ppg0, ppg1, ppg2], ambientSample: ambient, status: nil, frameType: frame.frameType))
            timeStampIndex += 1
        }
        
        return PpgData(timeStamp: frame.timeStamp, samples: ppgSamples)
    }
    
    private static func dataFromRawType6(frame: PmdDataFrame) throws -> PpgData {
        var ppgSamples = [PpgSample]()
        
        let sportId = TypeUtils.convertArrayToUnsignedInt64(frame.dataContent, offset: 0, size: TYPE_6_SAMPLE_SIZE_IN_BYTES)
        let samplesSize = frame.dataContent.count / TYPE_6_SAMPLE_SIZE_IN_BYTES
        let timeStamps = try PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp: frame.previousTimeStamp, frameTimeStamp: frame.timeStamp, samplesSize: UInt(samplesSize), sampleRate: frame.sampleRate)
        ppgSamples.append(
            PpgSample( timeStamp: timeStamps.first!, ppgDataSamples: [Int32(sportId)], ambientSample: nil, status: nil, frameType: frame.frameType)
        )
        
        return PpgData(timeStamp: timeStamps.first!, samples: ppgSamples)
    }
    
    private static func dataFromRawType9(frame: PmdDataFrame) throws -> PpgData {
        
        let step = TYPE_0_SAMPLE_SIZE_IN_BYTES
        let samplesSize = Int(Double(frame.dataContent.count) / Double(step * TYPE_0_CHANNELS_IN_SAMPLE))
        let timeStamps = try PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp: frame.previousTimeStamp, frameTimeStamp: frame.timeStamp, samplesSize: UInt(samplesSize), sampleRate: frame.sampleRate)
        var ppgSamples = [PpgSample]()
        var timeStampIndex = 0
        
        var offset = 0
        while offset < frame.dataContent.count {
            let numIntTs =
            frame.dataContent[offset..<(offset + TYPE_9_NUM_INTS_SIZE)].map(Int32.init)
            
            offset += TYPE_9_NUM_INTS_SIZE
            var channel1GainTs = [Int32]()
            for (index, value) in frame.dataContent[offset..<(offset + TYPE_9_CHANNEL_0_AND_1_SIZE)]
                .enumerated() {
                if (index % 2 == 0) {
                    channel1GainTs.append(Int32(value & 0x07))
                }
            }
            
            var channel2GainTs = [Int32]()
            for (index, value) in frame.dataContent[offset..<(offset + TYPE_9_CHANNEL_0_AND_1_SIZE)]
                .enumerated() {
                if (index % 2 == 1) {
                    channel2GainTs.append(Int32(value & 0x07))
                }
            }
            offset += TYPE_9_CHANNEL_0_AND_1_SIZE

            ppgSamples.append( PpgSample( timeStamp: timeStamps[timeStampIndex], ppgDataSamples: numIntTs, ambientSample: nil, status: nil, frameType: frame.frameType))
            ppgSamples.append( PpgSample( timeStamp: timeStamps[timeStampIndex], ppgDataSamples: channel1GainTs, ambientSample: nil, status: nil, frameType: frame.frameType))
            ppgSamples.append( PpgSample( timeStamp: timeStamps[timeStampIndex], ppgDataSamples: channel2GainTs, ambientSample: nil, status: nil, frameType: frame.frameType))
            timeStampIndex += 1
        }
        
        return PpgData(timeStamp: frame.timeStamp, samples: ppgSamples)
    }
    
    private static func dataFromCompressedType0(frame: PmdDataFrame) throws -> PpgData {
        let samples = Pmd.parseDeltaFramesToSamples(frame.dataContent, channels: TYPE_0_CHANNELS_IN_SAMPLE, resolution: TYPE_0_SAMPLE_SIZE_IN_BITS)
        let timeStamps = try PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp: frame.previousTimeStamp, frameTimeStamp: frame.timeStamp, samplesSize: UInt(samples.count), sampleRate: frame.sampleRate)
        
        var ppgSamples = [PpgSample]()
        for (index, sample) in samples.enumerated() {
            let ppg0:Int32 = sample[0]
            let ppg1:Int32 = sample[1]
            let ppg2:Int32 = sample[2]
            let ambient: Int32 = sample[3]
            ppgSamples.append( PpgSample( timeStamp: timeStamps[index], ppgDataSamples: [ppg0, ppg1, ppg2], ambientSample: ambient, status: nil, frameType: frame.frameType))
        }
        return PpgData(timeStamp: frame.timeStamp, samples: ppgSamples)
    }
    
    private static func dataFromCompressedType7(frame: PmdDataFrame) throws -> PpgData {
            let samples = Pmd.parseDeltaFramesToSamples(frame.dataContent, channels: TYPE_7_CHANNELS_IN_SAMPLE, resolution: TYPE_7_SAMPLE_SIZE_IN_BITS)
        let timeStamps = try PmdTimeStampUtils.getTimeStamps(previousFrameTimeStamp: frame.previousTimeStamp, frameTimeStamp: frame.timeStamp, samplesSize: UInt(samples.count), sampleRate: frame.sampleRate)
        var ppgSamplesFrameType7 = [PpgSample]()
        for (index, sample) in samples.enumerated() {
            let channelSamples = sample.map{ item in
                if (frame.factor != 1.0) {
                    return Int32(Float(item) * frame.factor)
                }
                else {
                    return item
                }
            }
            let status = Int32(sample[16] & 0xFFFFFF)
            ppgSamplesFrameType7.append( PpgSample( timeStamp: timeStamps[index], ppgDataSamples: channelSamples, ambientSample: nil, status: status, frameType: frame.frameType))
        }
        return PpgData(timeStamp: frame.timeStamp, samples: ppgSamplesFrameType7)
    }
    
    private static func dataFromCompressedType10(frame: PmdDataFrame) throws -> PpgData {
        
        let samples = Pmd.parseDeltaFramesToSamples(frame.dataContent, channels: TYPE_10_CHANNELS_IN_SAMPLE, resolution: TYPE_10_SAMPLE_SIZE_IN_BITS)

        let timeStamps = try PmdTimeStampUtils.getTimeStamps(
            previousFrameTimeStamp: frame.previousTimeStamp,
            frameTimeStamp: frame.timeStamp,
            samplesSize: UInt(samples.count),
            sampleRate: frame.sampleRate
        )
        var ppgSamples = [PpgSample]()
        var timeStampIndex = 0
        
        for (index, sample) in samples.enumerated() {
            
            let greenSamples = sample[0..<8].map { sample in
                if (frame.factor != Float(1.0)) {
                    Int32((Float(sample) * frame.factor))
                } else {
                    sample
                }
            }
            
            let redSamples = sample[8..<14].map { sample in
                if (frame.factor != Float(1.0)) {
                    Int32((Float(sample) * frame.factor))
                } else {
                    sample
                }
            }
            
            let irSamples = sample[14..<20].map { sample in
                if (frame.factor != Float(1.0)) {
                    Int32((Float(sample) * frame.factor))
                } else {
                    sample
                }
            }
            
            let ppgSample = PpgSample(
                timeStamp: timeStamps[index],
                ppgDataSamples: greenSamples + redSamples + irSamples,
                ambientSample: 0,
                status: sample[sample.endIndex - 1],
                frameType: frame.frameType
            )
            
            ppgSamples.append(ppgSample)
            timeStampIndex+=1
        }
        return PpgData(timeStamp: frame.timeStamp, samples: ppgSamples)
    }
}
